pkg/plugins/runtime/gateway/route_table_generator.go

```diff
@@ -110,3 +120,108 @@ func (r *RouteTableGenerator) GenerateHost(ctx xds_context.Context, info *Gatewa
 
 	return resources.Get(), nil
 }
+
+// retryRouteConfigurers returns the set of route configurers needed to implement the retry policy (if there is one).
+func retryRouteConfigurers(protocol core_mesh.Protocol, retry *core_mesh.RetryResource) []route.RouteConfigurer {
+	if retry == nil {
+		return nil
+	}
+
+	methodStrings := func(methods []mesh_proto.HttpMethod) []string {
+		var names []string
+		for _, m := range methods {
+			if m != mesh_proto.HttpMethod_NONE {
+				names = append(names, m.String())
+			}
+		}
+		return names
+	}
+
+	grpcConditionStrings := func(conditions []mesh_proto.Retry_Conf_Grpc_RetryOn) []string {
+		var names []string
+		for _, c := range conditions {
+			names = append(names, c.String())
+		}
+		return names
+	}
+
+	configurers := []route.RouteConfigurer{
+		route.RouteActionRetryDefault(protocol),
+	}
+
+	switch protocol {
+	case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2:
+		conf := retry.Spec.GetConf().GetHttp()
+		configurers = append(configurers,
+			route.RouteActionRetryOnStatus(conf.GetRetriableStatusCodes()...),
+			route.RouteActionRetryMethods(methodStrings(conf.GetRetriableMethods())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
+			route.RouteActionRetryCount(conf.GetNumRetries().GetValue()),
+			route.RouteActionRetryBackoff(
+				conf.GetBackOff().GetBaseInterval().AsDuration(),
+				conf.GetBackOff().GetMaxInterval().AsDuration()),
+		)
+	case core_mesh.ProtocolGRPC:
+		conf := retry.Spec.GetConf().GetGrpc()
+		configurers = append(configurers,
+			route.RouteActionRetryOnConditions(grpcConditionStrings(conf.GetRetryOn())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
```

Comment:
Might be a bit more readable if the common/non-protocol specific configurers are separated out

Reply:
It would, but the common fields are duplicated on each side of the Retry proto (i.e. in the "grpc" and "http" messages). So it would be complicated to flatten that.

Reply:
OK, the `conf` variable has a different type and it'd need an interface... probably not worth it